internal/components/setup/compose.go

```diff
@@ -17,3 +17,212 @@
 //
 
 package setup
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"regexp"
+	"syscall"
+	"time"
+
+	"github.com/apache/skywalking-infra-e2e/internal/config"
+	"github.com/apache/skywalking-infra-e2e/internal/constant"
+	"github.com/apache/skywalking-infra-e2e/internal/logger"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+
+	"github.com/testcontainers/testcontainers-go"
+)
+
+// ComposeSetup sets up environment according to e2e.yaml.
+func ComposeSetup(e2eConfig *config.E2EConfig) error {
+	composeConfigPath := e2eConfig.Setup.GetFile()
+	if composeConfigPath == "" {
+		return fmt.Errorf("no compose config file was provided")
+	}
+
+	// build docker client
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return err
+	}
+
+	// setup docker compose
+	composeFilePaths := []string{
+		composeConfigPath,
+	}
+	identifier := GetIdentity()
+	compose := testcontainers.NewLocalDockerCompose(composeFilePaths, identifier)
+	execError := compose.WithCommand([]string{"up", "-d"}).Invoke()
+	if execError.Error != nil {
+		return execError.Error
+	}
+
+	// record time now
+	timeNow := time.Now()
+	timeout := e2eConfig.Setup.Timeout
+	var waitTimeout time.Duration
+	if timeout <= 0 {
+		waitTimeout = constant.DefaultWaitTimeout
+	} else {
+		waitTimeout = time.Duration(timeout) * time.Second
+	}
+	logger.Log.Debugf("wait timeout is %d seconds", int(waitTimeout.Seconds()))
+
+	// find exported port and build env
+	for service, content := range compose.Services {
+		serviceConfig := content.(map[interface{}]interface{})
+		ports := serviceConfig["ports"]
+		if ports == nil {
+			continue
+		}
+		portList := ports.([]interface{})
+		container, err := findContainer(cli, fmt.Sprintf("%s_%s", identifier, getInstanceName(service)))
+		if err != nil {
+			return err
+		}
+		containerPorts := container.Ports
+
+		for inx := range portList {
+			for _, containerPort := range containerPorts {
+				if int(containerPort.PrivatePort) != portList[inx].(int) {
+					continue
+				}
+
+				// calculate max wait time
+				waitTimeout = NewTimeout(timeNow, waitTimeout)
+				timeNow = time.Now()
+
+				// wait port and export
```

Comment:
We don't need to wait and export, just exporting it is OK, the healthiness check should be done by `docker-compose.yml`, because the healthiness check should only make sense by the one who write `docker-compose.yml`, we can cover every possibility of how to check healthiness.

Reply:
Sure, but the `healthcheck` is only working on the service that has `depends_on`. If we need to use `webapp` service on `trigger` or `query`, it hasn't service dependent on this, so we may not access it. Do I miss anything?

Reply:
OK, it makes sense to me, can you reuse `compose.WithExposedService(service, port, wait.ForListeningPort("<port>/tcp"))` to replace `waitTCPPortStarted` so that we don't replicate the codes? Others generally look good to me, thanks again ❤️

Reply:
I have tried `compose.WithExposedService(service, port, wait.ForListeningPort("<port>/tcp"))` method, but currently the second parameter only supports the specified port, so we cannot use it. This error has already been reported and a new PR has been created. I will pay attention to the next release version of the `testcontainers-go` project. If this bug is fixed, I will use it.

Reply:
Thanks for the information, can you paste the related issue / PR here?

Reply:
Sorry I forget. here: https://github.com/testcontainers/testcontainers-go/pull/330

Reply:
> A new release has been pushed, fixing this. Hope it helps!

Wow, thx a lot. I will continue to work on this.